### PR TITLE
Refactor update

### PR DESCRIPTION
**Pull request by menjarleev**
*Issue #, if available:
#192 

*Description of changes:

* refactor main to support nn.Module like Model
* simplify training & test to single-GPU version
* As pytorch does not support sparse regularization(need to be verified), the regularizer does not work very well.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/dglke/__init__.py [hunk 1]
```diff
@@ -1,24 +0,0 @@
-# -*- coding: utf-8 -*-
```
Comment (classicsong) [inline python/dglke/__init__.py:1]: Why we remove this file?

File: python/dglke/data/dataloader/dataloader_generator.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                                persistent_workers=self.persistent_workers,
+                                num_workers=self.num_workers)
+        return EvalDataLoader(dataloader, self.metadata)
+
+# wrapper for dataloader to pass metadata into arguments
+class TrainDataLoader:
+    def __init__(self,
+                 dataloader,
+                 train_args=dict()):
+        self.dataloader = dataloader
```
Comment (classicsong) [inline python/dglke/data/dataloader/dataloader_generator.py:140]: Can you put some doc string here?

File: python/dglke/data/dataloader/dataloader_generator.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        self.data = train_args
```
Comment (classicsong) [inline python/dglke/data/dataloader/dataloader_generator.py:141]: Data is the arguments?

File: python/dglke/data/dataset/KGDataset.py [hunk 4]
```diff
@@ -117,7 +117,7 @@ def read_relation(self, relation_path):
 
         return relation2id, len(relation2id)
 
-    def read_triple(self, path, mode, skip_first_line=False, format=[0,1,2]):
+    def read_triple(self, path, mode, skip_first_line=False, format=[0,1,2], inverse_rel=False):
```
Comment (classicsong) [inline python/dglke/data/dataset/KGDataset.py:120]: Why we need inverse_rel?

File: python/dglke/data/dataset/partition_dataset.py [hunk 5]
```diff
@@ -0,0 +1,537 @@
...
+
+    def __getitem__(self, index):
+        return {'head': self.heads[index],
+                'rel': self.rels[index],
+                'tail': self.tails[index]}
+
+    def __len__(self):
+        return len(self.heads)
+
+class PartitionChunkDataset(th.utils.data.IterableDataset):
```
Comment (classicsong) [inline python/dglke/data/dataset/partition_dataset.py:501]: Any doc for this?

File: python/dglke/nn/modules/decoder.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+            else:
+                raise ValueError(f"{data['neg_type']} is not correct, choose from head, tail, both.")
+
+    def infer(self, head_emb, rel_emb, tail_emb):
+        pass
+
+    def evaluate(self, results, data, graph):
+        return self.metrics_evaluator.evaluate(results, data, graph)
+
+class AttHDecoder(BaseDecoder):
```
Comment (classicsong) [inline python/dglke/nn/modules/decoder.py:91]: Any docs?

File: python/dglke/nn/modules/decoder.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        raise NotImplementedError
+
+    def infer(self, encoded_data, data, gpu_id):
+        raise NotImplementedError
+
+    def get_loss(self, results):
+        return self.loss_generator.get_total_loss(results)
+
+
+class KGEDecoder(BaseDecoder):
```
Comment (classicsong) [inline python/dglke/nn/modules/decoder.py:36]: Docs here

File: python/dglke/nn/modules/model.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
+import torch as th
+import torch.nn as nn
+import os
+from itertools import chain
+from tqdm import trange, tqdm
+from utils import get_scalar
+import time
+
+class Model(nn.Module):
```
Comment (classicsong) [inline python/dglke/nn/modules/model.py:9]: KEModel can be a better name

**Review by classicsong — COMMENTED** [review]
Please add more docs

File: python/dglke/data/dataset/KGDataset.py (hunk 4 above)
Comment (menjarleev) [reply python/dglke/data/dataset/KGDataset.py:120]: Something trick for Hyperbolic Embedding. It treats head->rel->tail and tail->rel->head differently and auto-augment the data to have tail->inv_rel->head.

**Pull request merged**